### PR TITLE
feat: use trace-context "headers" in SNS and SQS single event message attributes for lambda transaction

### PR DESCRIPTION
**Pull request by trentm**
- If a SNS or SQS single event trigger to an instrumented Lambda
  function includes message attributes with the name "traceparent" (and
  "tracestate"), case-insensitive, then those are used to continue the
  trace. This was already being done for API Gateway event headers.
- Also, fix a bug in the capturing of SNS and SQS message attributes as
  "transaction.context.message.headers". Note that only "String"-type
  message attributes are captured. Binary-type attributes are base64
  encoded. I'm not sure of the value in capturing them.

Closes: #1831
Fixes: #2605
Refs: https://github.com/elastic/apm/pull/614

### Checklist

- [x] Implement code
- [x] Add tests
- [x] Add CHANGELOG.asciidoc entry

### CODE REVIEW

File: lib/lambda.js [hunk 1]
```diff
@@ -9,6 +9,32 @@ const path = require('path')
 let isFirstRun = true
 let gFaasId // Set on first invocation.
 
+// The trigger types for which we support special handling.
+// https://docs.aws.amazon.com/lambda/latest/dg/lambda-services.html
+const TRIGGER_GENERIC = 1
```
Comment (astorm) [inline lib/lambda.js:14]: I see we've replaced the smaller functions that identify an individual payload type (ex. `isApiGatewayEvent`) with this larger/more-complex function `triggerTypeFromEvent` function that identifies an individual payload type, and then we call `triggerTypeFromEvent` once its result value is passed around where it needs to be.  The main tradeoffs here seem to be 

- the perf. of multiple calls to the older small function vs. the higher memory use of keeping the constant-ish `TRIGGER_` variables around. 

- Swapping where the complexity goes :) 

No objections, just noting the change out loud.

Comment (astorm) [inline lib/lambda.js:14]: I see we've replaced the smaller functions that identify an individual payload type (ex. `isApiGatewayEvent`) with this larger/more-complex function `triggerTypeFromEvent` function that identifies an individual payload type, and then we call `triggerTypeFromEvent` once its result value is passed around where it needs to be.  The main tradeoffs here seem to be 

- the perf. of multiple calls to the older small function vs. the higher memory use of keeping the constant-ish `TRIGGER_` variables around. 

- Swapping where the complexity goes :) 

No objections, just noting the change out loud.

File: lib/lambda.js [hunk 2]
```diff
@@ -199,7 +221,10 @@ function setSqsData (agent, trans, event, context, faasId, isColdStart) {
   }
 
   if (agent._conf.captureHeaders) {
-    messageContext.headers = Object.assign({}, record.messageAttributes)
+    const headers = headersFromSqsMessageAttributes(record.messageAttributes)
+    if (Object.keys(headers).length > 0) {
```
Comment (astorm) [inline lib/lambda.js:225]: :+1: for `Object.keys(obj)` -- I'm pretty sure that's the fastest way to do this without needing to get into `for in` and `hasOwnProperty` territory. Unrelated to this PR/note-to-self -- a custom eslint linting rule that ensures we're using the more efficient Object.keys() for object looping/iteration (vs. one of the other myriad ways you can loop an object in javascript) would be a nice to have.

Comment (astorm) [inline lib/lambda.js:225]: :+1: for `Object.keys(obj)` -- I'm pretty sure that's the fastest way to do this without needing to get into `for in` and `hasOwnProperty` territory. Unrelated to this PR/note-to-self -- a custom eslint linting rule that ensures we're using the more efficient Object.keys() for object looping/iteration (vs. one of the other myriad ways you can loop an object in javascript) would be a nice to have.

File: lib/lambda.js [hunk 3]
```diff
@@ -559,10 +594,57 @@ function wrapNestedPath (agent, field, module) {
...
+
+// Reduce a SQS record message attributes object to a headers-like object of
+// string key/value-pairs. Some attribute types are ignored.
+// https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-message-metadata.html#message-attribute-data-types
+//
+// `record.messageAttributes` is of the form:
+//    { <attribute-name>: { dataType: <attr-type>, stringValue: <attr-value>, ... } }
+// For example:
+//    { traceparent: { dataType: 'String', stringValue: 'test-traceparent' } }
+function headersFromSqsMessageAttributes (messageAttributes) {
```
Comment (astorm) [inline lib/lambda.js:613]: 👍 I see a new `headersFromSqsMessageAttributes` that simplifies the data from amazon's strong-ishly typed JSON data into something that APM Server can handle. (I also see a separate function for SNS messages)

Comment (astorm) [inline lib/lambda.js:613]: 👍 I see a new `headersFromSqsMessageAttributes` that simplifies the data from amazon's strong-ishly typed JSON data into something that APM Server can handle. (I also see a separate function for SNS messages)

File: lib/lambda.js [hunk 4]
```diff
@@ -411,7 +410,12 @@ function elasticApmAwsLambda (agent) {
     }
 
     return function wrappedLambda (event, context, callback) {
-      log.trace({ awsRequestId: context && context.awsRequestId }, 'lambda: fn start')
+      if (!(event && context && typeof callback === 'function')) {
```
Comment (astorm) [inline lib/lambda.js:413]: 👍  for guarding against unknown datatypes here, per previous discussions on how things can go wrong without this.

Comment (astorm) [inline lib/lambda.js:413]: 👍  for guarding against unknown datatypes here, per previous discussions on how things can go wrong without this.

File: lib/lambda.js [hunk 5]
```diff
@@ -432,30 +436,61 @@ function elasticApmAwsLambda (agent) {
...
+          normedHeaders = lowerCaseObjectKeys(headersFromSqsMessageAttributes(record.messageAttributes))
         }
-        parentId = normalizedHeaders.traceparent ? normalizedHeaders.traceparent : normalizedHeaders['elastic-apm-traceparent']
-        tracestate = normalizedHeaders.tracestate
+      } else if (triggerType === TRIGGER_SNS_SINGLE_EVENT) {
+        const record = event.Records[0]
+        if (record.Sns && record.Sns.MessageAttributes) {
+          normedHeaders = lowerCaseObjectKeys(headersFromSnsMessageAttributes(record.Sns.MessageAttributes))
+        }
+      }
```
Comment (astorm) [inline lib/lambda.js:459]: Not required for this PR, but I'm genuinely unsure if we should mention the agent's  automatic handling of reading the trace context headers in a Lambda environment in our message queue doc, or if that would confusing things more https://github.com/elastic/apm-agent-nodejs/blob/main/docs/message-queues.asciidoc

Comment (astorm) [inline lib/lambda.js:459]: Not required for this PR, but I'm genuinely unsure if we should mention the agent's  automatic handling of reading the trace context headers in a Lambda environment in our message queue doc, or if that would confusing things more https://github.com/elastic/apm-agent-nodejs/blob/main/docs/message-queues.asciidoc

**Review by astorm — COMMENTED** [review]
Nothing objectionable here. It looks like _the thing to do_ and the approach seems reasonable.  Happy to approve once a changelog entry gets added.

I noticed this was still in draft state -- if there's a specific part you wanted feedback  on or  a part you think needs more extensive testing from a second set of eyes just let me know.  Otherwise 👍

File: lib/lambda.js (hunk 5 above)
Comment (trentm) [reply lib/lambda.js:459]: I think it is worthwhile to mention this behaviour in the *lambda.asciidoc*. Likely a reference in the message-queues doc would be worthwhile as well, I'd have to re-read that doc.

**Comment by trentm** [comment]
> I noticed this was still in draft state

Ah, thanks. I was going to pull out of draft after the changelog entry and merge from main.

**Comment by trentm** [comment]
Also there is still some discussion on the apm spec: https://github.com/elastic/apm/pull/614#discussion_r823436714
So I'll wait for that to settle before finalizing this.

**Comment by trentm** [comment]
Trying that failed "Integration Tests" again: https://apm-ci.elastic.co/job/apm-integration-tests-selector-mbp/job/main/2446/

**Review by astorm — APPROVED** [review]

**Pull request merged**